runtime/src/deterministic.rs

```diff
@@ -387,86 +390,234 @@ pub struct Executor {
-        let auditor = Arc::new(Auditor::default());
-        let start_time = UNIX_EPOCH;
-        let deadline = cfg
-            .timeout
-            .map(|timeout| start_time.checked_add(timeout).expect("timeout overflowed"));
-        let (signaler, signal) = Signaler::new();
-        let storage = MeteredStorage::new(
-            AuditedStorage::new(MemStorage::default(), auditor.clone()),
-            runtime_registry,
-        );
-        let executor = Arc::new(Self {
-            registry: Mutex::new(registry),
-            cycle: cfg.cycle,
-            deadline,
-            metrics: metrics.clone(),
-            auditor: auditor.clone(),
-            rng: Mutex::new(StdRng::seed_from_u64(cfg.seed)),
-            time: Mutex::new(start_time),
-            tasks: Arc::new(Tasks::new()),
-            sleeping: Mutex::new(BinaryHeap::new()),
-            partitions: Mutex::new(HashMap::new()),
-            signaler: Mutex::new(signaler),
-            signal,
-            finished: Mutex::new(false),
-            recovered: Mutex::new(false),
-        });
-        (
-            Runner {
-                executor: executor.clone(),
-            },
-            Context {
-                label: String::new(),
-                spawned: false,
-                executor,
-                networking: Arc::new(Networking::new(metrics, auditor.clone())),
-                storage,
-            },
-            auditor,
-        )
+        Runner {
+            state: RunnerState::NoContext(cfg),
+        }
     }
 
     /// Initialize a new `deterministic` runtime with the default configuration
     /// and the provided seed.
-    pub fn seeded(seed: u64) -> (Runner, Context, Arc<Auditor>) {
+    pub fn seeded(seed: u64) -> Self {
         let cfg = Config {
             seed,
             ..Config::default()
         };
-        Self::init(cfg)
+        Self::new(cfg)
     }
 
     /// Initialize a new `deterministic` runtime with the default configuration
     /// but exit after the given timeout.
-    pub fn timed(timeout: Duration) -> (Runner, Context, Arc<Auditor>) {
+    pub fn timed(timeout: Duration) -> Self {
         let cfg = Config {
             timeout: Some(timeout),
             ..Config::default()
         };
-        Self::init(cfg)
+        Self::new(cfg)
     }
+}
 
-    /// Initialize a new `deterministic` runtime with the default configuration.
-    // We'd love to implement the trait but we can't because of the return type.
-    #[allow(clippy::should_implement_trait)]
-    pub fn default() -> (Runner, Context, Arc<Auditor>) {
-        Self::init(Config::default())
+impl Default for Runner {
+    fn default() -> Self {
+        Self::new(Config::default())
+    }
+}
+
+impl crate::Runner for Runner {
+    type Context = Context;
+
+    fn start<F, Fut>(mut self, f: F) -> Fut::Output
+    where
+        F: FnOnce(Self::Context) -> Fut,
+        Fut: Future,
+    {
+        match self.state {
```

Comment:
nit: to avoid nesting/recursion, recommend populating context as a var before entering the actual logic of `start`